### PR TITLE
Better validation of domains when creating TensorDomain

### PR DESCRIPTION
**Pull request by naoyam**
This is for #225 

A `TensorDomain` is created with a list of IterDomains representing a root domain, and optionally another list of IterDomains for an rfactor domain and also one more list for the leaf domain.

Currently, we do some error check using `IterVisitor::getInputsTo`, but that's not always sufficient. We need to make sure that the root domain and the other domains represent the same logical space without any overlap. 

For example:

```
IterDomain* id0 = IterDomainBuilder()....build();
IterDomain* id1, id2 = IterDomain::split(id0, 4);
```

Here, it'd be valid to have a root and leaf domain as follows:

```
Root: id0
Leaf: id1, id2
```

Suppose `id2` is further split as:

```
IterDomain* id3, id4 = IterDomain::split(id2, 5);
```
Then, this is invalid:

```
Root: id0
Leaf: id1, id2, id3, id4
```

`id3` and `id4` are disjoint, but they overlap with `id2`. To represent the original space using the new leaf IDs, `id2` shouldn't be included, however, the current error check won't detect these kinds of invalid domains.

This PR replaces the existing error check with more complete analysis. As long as all IDs are non-symbolic, it should be able to determine whether a list of initial IDs and another list of derived IDs represent the same space. It also checks if the derived IDs are disjoint.

Since the error check is always used when TensorDomain is created, I consider it well tested, so I just added one unit test for testing some corner cases.

### CODE REVIEW

**Review by liqiangxl — APPROVED** [review]
LGTM.
Do we have a case cocver symbolic IDs?

**Comment by naoyam** [comment]
> LGTM. Do we have a case cocver symbolic IDs?

Thanks. That's a good point. I'll add one.

File: csrc/ir_utils.cpp [hunk 1]
```diff
@@ -852,5 +852,94 @@ std::vector<TensorView*> getTVsWithDynamicTransform(Fusion* fusion) {
...
+        std::none_of(
+            expr->inputs().begin(),
+            expr->inputs().end(),
+            [&](Val* input_val) {
+              return derived_domain_.find(input_val) != derived_domain_.end();
+            }),
+        "Invalid derived domain due to dependent expr: ",
+        expr->toString(),
+        ". Derived domain: ",
+        toDelimitedString(derived_domain_));
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/ir_utils.cpp]: No, I don't think so. I tried with the last case of the test (0a7c92df8f2048c67c7434c1a0c9be47b8eecd6). An error was reported but it was at the end of the constructor, not these new assertions. The error was due to the final frontier node set is different from the derived domain. I think it's more informative to detect the error earlier at this point.

File: csrc/ir_utils.cpp [hunk 2]
```diff
@@ -852,5 +852,94 @@ std::vector<TensorView*> getTVsWithDynamicTransform(Fusion* fusion) {
...
+          toDelimitedString(derived_domain));
+    }
+  };
+
+  std::vector<Statement*> next(Val* v) override {
+    if (initial_domain_.find(v) != initial_domain_.end()) {
+      return std::vector<Statement*>();
+    }
+    return IterVisitor::next(v);
+  }
```
[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/ir_utils.cpp:903]: I think so. I initially tried to use some of the existing routines, but couldn't make it work, so I created a new class inheriting IterVisitor, and then eventually I was able to simplify the analysis, and now it looks like creating the class isn't worth it. Let me rewrite it.

[the remark this answers is not in the document]
Comment (naoyam) [reply csrc/ir_utils.cpp:903]: Decided to keep using a custom subclass, but switched to use `traverseBetween`. Thanks for the suggestion.

**Review by zasdfgbnm — APPROVED** [review]

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build
